File: sandbox/js/envelope_endpoint_mock.mjs
import { isEnvelope } from "../../packages/js/envelope.mjs";

export class MockEnvelopeEndpoint {
  constructor() {
    this.pending = [];
    this.waiters = [];
  }

  send(envelope) {
    if (!isEnvelope(envelope)) {
      throw new Error("Can only send JobPing envelopes");
    }

    const waiterIndex = this.waiters.findIndex((waiter) => waiter.matches(envelope));
    if (waiterIndex !== -1) {
      const [waiter] = this.waiters.splice(waiterIndex, 1);
      waiter.resolve(envelope);
      return;
    }

    this.pending.push(envelope);
  }

  recv({ jobId, type, timeoutMs = 1000 } = {}) {
    const pendingIndex = this.pending.findIndex((envelope) =>
      this.matchesEnvelope(envelope, jobId, type),
    );

    if (pendingIndex !== -1) {
      const [envelope] = this.pending.splice(pendingIndex, 1);
      return Promise.resolve(envelope);
    }

    return new Promise((resolve, reject) => {
      const waiter = {
        matches: (envelope) => this.matchesEnvelope(envelope, jobId, type),
        resolve,
      };
      const timer = setTimeout(() => {
        const waiterIndex = this.waiters.indexOf(waiter);
        if (waiterIndex !== -1) {
          this.waiters.splice(waiterIndex, 1);
        }
        reject(new Error("Timed out waiting for JobPing envelope"));
      }, timeoutMs);

      waiter.resolve = (envelope) => {
        clearTimeout(timer);
        resolve(envelope);
      };
      this.waiters.push(waiter);
    });
  }

  size() {
    return {
      pending: this.pending.length,
      waiters: this.waiters.length,
    };
  }

  matchesEnvelope(envelope, jobId, type) {
    return (
      isEnvelope(envelope) &&
      (jobId === undefined || envelope.job_id === jobId) &&
      (type === undefined || envelope.type === type)
    );
  }
}

File: sandbox/python/jobping_sandbox/envelope_endpoint_mock.py
from __future__ import annotations

import asyncio
from dataclasses import dataclass

from jobping.envelope import EnvelopeType, JobPingEnvelope, is_envelope


@dataclass
class _Waiter:
    job_id: str | None
    type: EnvelopeType | None
    future: asyncio.Future[JobPingEnvelope]


class MockEnvelopeEndpoint:
    """In-memory endpoint for envelope send/recv tests."""

    def __init__(self) -> None:
        self._pending: list[JobPingEnvelope] = []
        self._waiters: list[_Waiter] = []

    def send(self, envelope: JobPingEnvelope) -> None:
        if not is_envelope(envelope):
            raise ValueError("Can only send JobPing envelopes")

        for index, waiter in enumerate(self._waiters):
            if self._matches(envelope, waiter.job_id, waiter.type):
                self._waiters.pop(index)
                waiter.future.set_result(envelope)
                return

        self._pending.append(envelope)

    async def recv(
        self,
        *,
        job_id: str | None = None,
        type: EnvelopeType | None = None,
        timeout: float = 1.0,
    ) -> JobPingEnvelope:
        for index, envelope in enumerate(self._pending):
            if self._matches(envelope, job_id, type):
                return self._pending.pop(index)

        loop = asyncio.get_running_loop()
        waiter = _Waiter(
            job_id=job_id,
            type=type,
            future=loop.create_future(),
        )
        self._waiters.append(waiter)

        try:
            return await asyncio.wait_for(waiter.future, timeout=timeout)
        except TimeoutError:
            if waiter in self._waiters:
                self._waiters.remove(waiter)
            raise

    def size(self) -> dict[str, int]:
        return {
            "pending": len(self._pending),
            "waiters": len(self._waiters),
        }

    def _matches(
        self,
        envelope: JobPingEnvelope,
        job_id: str | None,
        type: EnvelopeType | None,
    ) -> bool:
        return (
            is_envelope(envelope)
            and (job_id is None or envelope["job_id"] == job_id)
            and (type is None or envelope["type"] == type)
        )
